lib/src/rules/avoid_runtimeType_toString.dart

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Avoid calling toString() on runtimeType.';
+
+const _details = r'''
+
+Calling `toString` on a runtime type is a non-trivial operation that can
+negatively impact performance. It's better to avoid it.
+
+**BAD:**
+```
+class A {
+  String toString() => '$runtimeType()';
+}
+```
+
+**GOOD:**
+```
+class A {
+  String toString() => 'A()';
+}
+```
+
+This lint has some exceptions where performance is not a problem or where real
+type information is more important than performance:
+
+* in assertion
+* in throw expressions
+* in catch clauses
+* in mixin declaration
+* in abstract class
+
+''';
+
+class AvoidRuntimeTypeToString extends LintRule implements NodeLintRule {
+  AvoidRuntimeTypeToString()
+      : super(
+            name: 'avoid_runtimeType_toString',
```

Comment:
In internal discussions I think we've decided to stop using prefixes like "avoid_" or "prefer_" when naming lints (and are thinking about what it would take to rename existing lints). Phil can confirm.

Reply:
Yeah, where possible it's nice to avoid `avoid` and `prefer`,  etc. in the name moving forward.  `runtimeType_toString` might be a little vague though (maybe prefix w/ `no_`?)